lib/wallet/src/Cardano/Wallet/Write/Tx.hs

```diff
@@ -454,6 +524,65 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
 downcastTxOut (TxOutInRecentEra addr val (Alonzo.DatumHash dh) Nothing)
     = Right $ Alonzo.TxOut addr val (SJust dh)
 
+--
+-- MinimumUTxO
+--
+
+-- |
+--
+-- prop> forall out.
+-- prop> let
+-- prop>     c = computeMinimumCoinForUTxO out
+-- prop> in
+-- prop>     forall c' > c. satisfiesMinimumUTxO (out { coin = c' })
+--
+-- NOTE:
+computeMinimumCoinForTxOut
+    :: forall era. RecentEra era -- FIXME [ADP-2353] Replace 'RecentEra' with 'IsRecentEra'
+    -> Core.PParams (ShelleyLedgerEra era)
+    -> TxOut (ShelleyLedgerEra era)
+    -> Coin
+computeMinimumCoinForTxOut era pp out = withCLIConstraint era $
+    evaluateMinLovelaceOutput pp (withMaxLengthSerializedCoin out)
+  where
+    withMaxLengthSerializedCoin
+        :: TxOut (ShelleyLedgerEra era)
+        -> TxOut (ShelleyLedgerEra era)
+    withMaxLengthSerializedCoin =
+        withStandardCryptoConstraint era $
+            modifyTxOutValue era $ modifyCoin (const $ toLedger txOutMaxCoin)
+
+isBelowMinimumCoinForTxOut
+    :: forall era. RecentEra era -- FIXME [ADP-2353] Replace 'RecentEra' with 'IsRecentEra'
+    -> Core.PParams (ShelleyLedgerEra era)
+    -> TxOut (ShelleyLedgerEra era)
+    -> Either Coin ()
+isBelowMinimumCoinForTxOut era pp out = withCLIConstraint era $
+    let
+        requiredMin = evaluateMinLovelaceOutput pp out
+        actualCoin = getCoin era out
+    in
+        if actualCoin >= requiredMin
+        then Right ()
+        else Left $ requiredMin <-> actualCoin
```

Comment:
🤔 First I didn't want to add heavy types (`expected: , actual:`) because it wasn't needed yet and returned a `Bool`. Then I needed some more details in the quickcheck property counterexample (currently the only caller of this function) and added the `shortfall`.

Reply:
Hmm.

Given that `evaluateMinLovelaceOutput` does not reach a fixed point, what should the caller do with the returned `Coin` deficit? It's not enough to simply compute `original <> deficit`, as this may cause an increase in the serialised length of the ada quantity, and then `evaluateMinLovelaceOutput` would presumably return an even higher value.

If the specified value is below the minimum, then why not return the result of `computeMinimumCoinForTxOut`? At least this way, the caller can choose any value greater than or equal to the returned value, and be confident that it'll work.

Returning a minimum that can always be safely increased is actually the behaviour of our existing API endpoints: I think it would be nice to preserve this behaviour!

To make the purpose of the returned `Coin` value more obvious, I would propose embedding it in a newtype:
```hs
newtype SafeMinimum a = SafeMinimum {getSafeMinimum :: a}
```
Then the return type of the function could be:
```hs
Either (SafeMinimum Coin) ()
```

Reply:
> Given that evaluateMinLovelaceOutput does not reach a fixed point

👍 Good point, I forgot about that here. 

Perhaps I'll leave the function returning `Bool` with a TODO to add a more descriptive error for the sake of 1) taking over responsibility from the checks in coin-selection 2) providing counterexample details in the property tests 🤔

Reply:
69a9f4e43f7f0e00113f22f1f71c0a5158e1dde9